### PR TITLE
Implemented the Telles quadrature formulas and their test

### PR DESCRIPTION
**Pull request by nicola-giuliani**

### CODE REVIEW

File: include/deal.II/base/quadrature_lib.h [hunk 1]
```diff
@@ -439,6 +439,71 @@ class QSorted : public Quadrature<dim>
...
 };
 
+/**
+ * Telles quadrature of arbitrary order.
+ *
+ * The coefficients of these quadrature rules are computed using
+ * a non linear variable change starting from a Gauss-Legendre
+ * quadrature formula.
+ * This is done using a cubic polynomial,
+ * n = a x^3 + b x^2 + c x + d
```
Comment (tamiko) [inline include/deal.II/base/quadrature_lib.h]: Can you set formulas in math mode please? (Follows latex syntax), $$ for inline formulas,

File: include/deal.II/base/quadrature_lib.h [hunk 2]
```diff
@@ -439,6 +439,71 @@ class QSorted : public Quadrature<dim>
...
+ *
+ * We start from a Gauss Quadrature Formula with arbitrary
+ * function. Then we apply the cubic variable change.
+ * In his paper Telles applies the transformation on the reference
+ * cell [-1, 1] getting
+ *
+ * n(1) = 1
+ * n(-1) = -1
+ * dn/dx = 0 at x = x_0
+ * d2n/dx2 = 0 at x = x_0
```
Comment (tamiko) [inline include/deal.II/base/quadrature_lib.h]:
```
@f{align*}
  ...
@f}
```
here

File: include/deal.II/base/quadrature_lib.h [hunk 3]
```diff
@@ -439,6 +439,71 @@ class QSorted : public Quadrature<dim>
...
+ *
+ * We get
+ * a = 1/q,
+ * b = -3gamma_bar/q,
+ * c = 3gamma_bar/q,
+ * d = -b,
+ * with
+ * eta_star=eta_bar^2 - 1;
+ * gamma_bar= nthroot( eta_bar*eta_star + abs(eta_star) ,3) + nthroot(eta_bar*eta_star - abs(eta_star),3)
```
Comment (tamiko) [inline include/deal.II/base/quadrature_lib.h]: Please break overlong lines

File: include/deal.II/base/quadrature_lib.h [hunk 4]
```diff
@@ -439,6 +439,71 @@ class QSorted : public Quadrature<dim>
...
+ * where $x_0$ is given at construction time, and is the location of the
+ * singularity $x_0$, and f(x) is a smooth non singular function.
+ *
+ * Singular quadrature formula are rather expensive, nevertheless Telles'
+ * quadrature formula are much easier to compute with respect to other singular
+ * integration techniques as Lachat-Watson.
+ *
+ * The weights and functions for Gauss Legendre formula have been tabulated up
+ * to order 12.
+ *
```
Comment (tamiko) [inline include/deal.II/base/quadrature_lib.h]: Feel free to add a line
```
 @author Nicola Giulani, 2015 
```
here.

File: include/deal.II/base/quadrature_lib.h [hunk 5]
```diff
@@ -439,6 +439,71 @@ class QSorted : public Quadrature<dim>
...
+template <int dim>
+class QTelles: public Quadrature<dim>
+{
+public:
+    /**
+      * In the constructor we can specify the order and the singular point,
+      * in this case we will map Gauss Legendre quadrature formulas following Telles'
+      * rule.
+    **/
+    QTelles (const unsigned int n, const Point<dim> singularity);
```
Comment (tjhei) [inline include/deal.II/base/quadrature_lib.h]: pass point by reference

File: include/deal.II/base/quadrature_lib.h [hunk 6]
```diff
@@ -439,6 +439,71 @@ class QSorted : public Quadrature<dim>
...
+    /**
+      * In the second constructor we can specify the quadrature formula to be mapped
+      * using Telles' rule, and the singular point.
+    **/
+
+    QTelles (const Quadrature<1> base_quad, const Point<dim> singularity);
```
Comment (tjhei) [inline include/deal.II/base/quadrature_lib.h]: make both arguments a const reference

File: source/base/quadrature_lib.cc [hunk 7]
```diff
@@ -1014,6 +1014,116 @@ QWeddle<dim>::QWeddle ()
...
+          quadrature_points.resize(quadrature_points_dummy.size()-1);
+          weights.resize(weights_dummy.size()-1);
+          for(unsigned int d=0; d<quadrature_points.size()-1; ++d)
+          {
+              quadrature_points[d]=quadrature_points_dummy[d];
+              weights[d]=weights_dummy[d];
+          }
+      }
+
+      if((eta_bar * eta_star + std::abs(eta_star))==0 || (eta_bar * eta_star - std::abs(eta_star))==0)
```
Comment (tjhei) [inline source/base/quadrature_lib.cc]: What is this check for? A floating point comparison with ==0 doesn't make a lot of sense (will rarely happen). You probably need a check with an epsilon here.

Comment (bangerth) [inline source/base/quadrature_lib.cc]: The condition also evaluates to `eta_star==0` :-)

**Comment by bangerth** [comment]
Also (and importantly): Welcome, @nicola-giuliani , to the deal.II project. If you wouldn't mind making the few changes commented on by @tjhei and @tamiko , we'd be very glad to merge your patch!

Just drop a note here once you have updated your patch!

File: include/deal.II/base/quadrature_lib.h (hunk 6 above)
Comment (tamiko) [inline include/deal.II/base/quadrature_lib.h]: Can you change the order of the two constructors? What about the following descriptions?
"A constructor that takes a quadrature formula and a singular point as argument. The quadrature formula will be mapped using Telles' rule. Make sure that the order of the quadrature rule is appropriate for the singularity in question."

"A variant of above constructor that takes as parameters the order @p n and location of a singularity. A Gauss Legendre quadrature of order n will be used."

**Comment by tamiko** [comment]
Very nice!

Please also indent the code with astyle and add a note to <code>changes.h</code>.

We can go through the individual comments together tomorrow if you wish.

File: source/base/quadrature_lib.cc [hunk 8]
```diff
@@ -1014,6 +1014,116 @@ QWeddle<dim>::QWeddle ()
...
+  * We explicitly implement the Telles' variable change if dim == 1.
+  **/
+      Quadrature<1>(base_quad)
+  {
+      /**
+      * We define all the constants to be used in the implementation of
+      * Telles' rule
+      **/
+      double eta_bar = singularity[0] * 2 - 1;
+      double eta_star = eta_bar * eta_bar - 1;
```
Comment (tamiko) [inline source/base/quadrature_lib.cc]: What about <code>const double eta_bar</code> and <code>const double eta_star</code>?

File: source/base/quadrature_lib.cc [hunk 9]
```diff
@@ -1014,6 +1014,116 @@ QWeddle<dim>::QWeddle ()
...
+      double gamma_bar;
+
+      std::vector<Point<1>> quadrature_points_dummy(quadrature_points.size());
+      std::vector<double> weights_dummy(weights.size());
+      unsigned int cont=0;
+      double toll=1e-5;
```
Comment (tamiko) [inline source/base/quadrature_lib.cc]: "tolleranza" <code>double toll</code> -> <code>const double tol</code>

We should probably choose something more generic than <code>1e-5</code> here.

File: tests/codim_one/integrate_one_over_r_telles.cc [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+
+
+  std::vector<Point<2> > vertices = FE_Q<2>(1).get_unit_support_points();
+
+  for (unsigned int m=1; m<7; ++m)
+    {
+      deallog << " =========Quadrature Order: " << m
+              << " =============================== " << endl;
+      deallog << " ============================================================ " << endl;
+      // for (unsigned int index=0; index<4; ++index)
```
Comment (tamiko) [inline tests/codim_one/integrate_one_over_r_telles.cc]: Can you remove dead code, please?

File: tests/codim_one/integrate_one_over_r_telles.cc [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+//      end proc;
+//      Digits := 20;
+//      for i from 3 to 3 do
+//         for n from 0 to 5 do
+//          for m from 0 to 5 do
+//               v[i+1][n+1][m+1] = sing_int(i, n, m);
+//            end do;
+//         end do;
+//      end do;
+//      C(v)
```
Comment (tamiko) [inline tests/codim_one/integrate_one_over_r_telles.cc]: Very nice! Good documentation style.

**Comment by nicola-giuliani** [comment]
I've found a bug in my implementation, I'm removing it right now. I will commit again as soon as everything works fine.

**Comment by nicola-giuliani** [comment]
I've removed the bug in my implementation, now it seems to work properly.

**Comment by tamiko** [comment]
Very nice. I think now all comments are addressed. @tjhei @bangerth can one of you please have a final look and merge :-]

File: tests/codim_one/integrate_one_over_r_telles.cc [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+    {
+      deallog << " =========Quadrature Order: " << m
+              << " =============================== " << endl;
+      deallog << " ============================================================ " << endl;
+      unsigned int index=0;
+        {
+          deallog << " ===============Vertex: " << vertices[index]
+                  << " ============================= " << endl;
+          QTelles<2> quad(6, vertices[index]);
+          QGaussOneOverR<2> quad2(4, vertices[index]);
```
Comment (luca-heltai) [inline tests/codim_one/integrate_one_over_r_telles.cc]: The quadrature order is never used. You are always checking Telles of order 6, and OneOverR of order 4. Can you put m in here?

**Comment by nicola-giuliani** [comment]
I've corrected the test and its corresponding output.

File: include/deal.II/base/quadrature_lib.h [hunk 13]
```diff
@@ -439,6 +439,77 @@ class QSorted : public Quadrature<dim>
                   const std::pair<double, Point<dim> > &b);
 };
 
+/**
+ * Telles quadrature of arbitrary order.
+ *
+ * The coefficients of these quadrature rules are computed using
+ * a non linear variable change starting from a Gauss-Legendre
```
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h]: variable change -> change of variables

File: include/deal.II/base/quadrature_lib.h [hunk 14]
```diff
@@ -439,6 +439,77 @@ class QSorted : public Quadrature<dim>
...
+ * n(1) = 1
+ * n(-1) = -1
+ * dn/dx = 0 at x = x_0
+ * d2n/dx2 = 0 at x = x_0
+ * @f}
+ * We get
+ * @f{align*}
+ * a = 1/q,
+ * b = -3gamma_bar/q,
+ * c = 3gamma_bar/q,
```
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h]: This won't look good in latex. You'll need to write it as \bar\gamma or similar.

File: include/deal.II/base/quadrature_lib.h [hunk 15]
```diff
@@ -439,6 +439,77 @@ class QSorted : public Quadrature<dim>
...
+ * d = -b,
+ * @f}
+ * with
+ * @f{align*}
+ * eta_star=eta_bar^2 - 1;
+ * gamma_bar = nthroot( eta_bar*eta_star + abs(eta_star) ,3) +
+ * nthroot(eta_bar*eta_star - abs(eta_star),3) + eta_bar;
+ * q=((gamma-gamma_bar).^3 + gamma_bar*(gamma_bar^2+3))/(1+3*gamma_bar^2);
```
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h]: similarly with these formulas.

File: include/deal.II/base/quadrature_lib.h [hunk 16]
```diff
@@ -439,6 +439,77 @@ class QSorted : public Quadrature<dim>
...
+ * a non linear variable change starting from a Gauss-Legendre
+ * quadrature formula.
+ * This is done using a cubic polynomial,
+ * $n = a x^3 + b x^2 + c x + d$
+ * in order to integrate
+ * a singular integral, with singularity at a given point x_0.
+ *
+ * We start from a Gauss Quadrature Formula with arbitrary
+ * function. Then we apply the cubic variable change.
+ * In his paper Telles applies the transformation on the reference
```
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h]: Can you place a full reference to that paper somewhere?

File: doc/news/changes.h [hunk 17]
```diff
@@ -402,9 +402,16 @@ inconvenience this causes.
...
 
 <ol>
-  <li> New: A function to get a map with all vertices at boundaries
-  has been added at GridTools::get_all_vertices_at_boundary.
-  This function will return a map which can be used in functions
+  <li> New: There is now a new quadrature formula in quadrature_lib. It is
+  now possible to use Telles' quadrature rules to integrate singular
+  integrals
+  <br>
+  (Nicola Giuliani, 2015/04/01)
```
Comment (bangerth) [inline doc/news/changes.h]: Please reference `QTelles` somewhere. This way doxygen automatically links to it for people to click through for the full description.

**Comment by bangerth** [comment]
Nice job! Just the few documentation comments to take care of. Can you add a couple of sentences to the documentation about what happens if `dim>1`? It's implemented but not described.

**Comment by nicola-giuliani** [comment]
Dear Wolfgang, thank you for the useful corrections. I've tried to improve the math formula. Can you tell me how to put a reference inside the deal documentation?

**Comment by bangerth** [comment]
Nothing fancy necessary -- just something like
```
   X. Telles: Quadrature rules for singular kernels. Journal of Improbable Research, vol. 24, 2015.
```

**Comment by nicola-giuliani** [comment]
Thank you Wolfgang, now I've added the reference to the original paper by Telles. Let me know if the documentation looks better now.

**Comment by bangerth** [comment]
Yes, looks good, many thanks! Last step: can you squash the 14 commits into one and push it to github again?

The commands should be something like
```
  git checkout qtelles
  git rebase -i master
```
then mark all but the first patch as "sqash", and finally
```
  git push -f origin qtellese
```
There is some potential for corrupting all of your data in these steps, so better have Luca or someone with git experience look over your shoulder if you don't feel comfortable doing it by yourself the first time.

**Comment by nicola-giuliani** [comment]
Thank you Wolfgang for the advices. I should have squash all my commits into one.

**Pull request merged**